### PR TITLE
Refactor SCF

### PR DESCRIPTION
**Pull request by antoine-levitt**
Merge the logic in damped and nlsolve

### CODE REVIEW

**Comment by antoine-levitt** [comment]
OK so I've got everything I wanted in there. Of course on silicon there's no real point in testing these things because it converges very quickly anyway... we should find ourselves a tricky case.

**Comment by mfherbst** [comment]
Looks generally quite good. I'll get into merging this and #8 next week.

**Comment by antoine-levitt** [comment]
OK ! No hurry. The jacobian with scaling=0.5 is still not symmetric, so something is going wrong.

**Comment by antoine-levitt** [comment]
That matrix is pretty interesting, although it possibly just means that my Ecut is too small. Should optimize that to access larger Ecuts.

**Comment by mfherbst** [comment]
If you are using Ecut=3, than that's really on the low end. How about we split this PR into two and first merge the refactoring bit and then the jacobian can go in later.

**Comment by antoine-levitt** [comment]
Sure! My git skills are not up to that, but I'll read about it and do that when I get the chance.

**Comment by mfherbst** [comment]
No worries, I'm happy to do that as well if needed :smile:. I guess: Whoever gets there first.

**Comment by antoine-levitt** [comment]
Done, I think. There are test failures but they appear unrelated?

**Comment by mfherbst** [comment]
Great. Good to know. I'm a bit busy at the moment, but I'll hopefully get to in the upcoming days.

**Comment by mfherbst** [comment]
Hmm I'm a bit confused by the current history. Is there any reason you did a merge and not a rebase?

**Comment by mfherbst** [comment]
Ok, I did some rebasing ... now it looks better to me.

**Comment by mfherbst** [comment]
By the way ... the old jacobian stuff can now be found in the branch compute_jacobian_scf if you want to continue working on it. It's based on the current tip of the refactor_scf.

**Comment by antoine-levitt** [comment]
Hm, essentially that I do not know the difference between merge and rebase... this branch stuff is new for me, usually I just work on master; I will read up on that. Thanks for keeping the jacobian stuff! I just copied the file somewhere :-p

**Comment by mfherbst** [comment]
We can discuss when I'm back. There are about 10 Qudrizillion ways to do git braching and so on. As usual it boils down to: One has to agree. I prefer to keep merging to a minimum, using rebase instead. Actually a pretty good video, which helped me a lot with understanding branching is https://www.youtube.com/watch?v=3m7BgIvC-uQ.

**Comment by antoine-levitt** [comment]
OK, I'll rebase in the future

File: src/core/scf.jl [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                 prec=lobpcg_prec, tol=lobpcg_tol)
+    @assert res.converged
+    Psi .= res.X
+    occupation = compute_occupation(ham.basis, res.λ, res.X)
+    ρ_new = compute_density(pw, res.X, occupation)
+end
+
+# Scaling is from 0 to 1. 0 is density mixing, 1 is "potential mixing"
+# (at least, Hartree potential mixing). 1/2 results in a symmetric
+# Jacobian of the SCF mapping (when there is no exchange-correlation)
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/core/scf.jl]: Also, I'm not sure I understand how this works. Can you explain?

File: src/core/scf_solvers.jl [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+function scf_nlsolve_solver(m)
+    function fp_solver(f, x0, tol, max_iter)
+        res = nlsolve(x -> f(x) - x, x0, method=:anderson, m=m, xtol=tol,
+                      ftol=0.0, show_trace=true, iterations=max_iter)
+        (sol=res.zero, converged=converged(res))
+    end
+    fp_solver
+end
+function scf_damping_solver(β)
```
Comment (mfherbst) [inline src/core/scf_solvers.jl]: Docstring: What is `β`?

File: src/core/scf_solvers.jl [hunk 3]
```diff
@@ -0,0 +1,133 @@
+# these provide fixed-point solvers that can be passed to scf()
+
+# the fp_solver function must accept being called like fp_solver(f, x0, tol,
+# maxiter), where f(x) is the fixed-point map. It must return an
+# object supporting res.sol and res.converged
+
+using NLsolve
+
+function scf_nlsolve_solver(m)
```
Comment (mfherbst) [inline src/core/scf_solvers.jl]: Docstring: What is `m`?

Comment (mfherbst) [reply src/core/scf_solvers.jl]: Maybe offer to tinker with even more parameters of nlsolve?

File: src/core/scf_solvers.jl [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        (sol=x, converged=converged)
+    end
+    fp_solver
+end
+
+
+
+## Basic versions of anderson mixing and CROP algorithm
+
+function anderson(g,x0,m::Int,niter::Int,eps::Real,warming=0)
```
Comment (mfherbst) [inline src/core/scf_solvers.jl]: Again I use the python convention: Space after the `,`.

File: src/core/scf_solvers.jl [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    @assert length(size(x0)) == 1 #1D array input
```
Comment (mfherbst) [inline src/core/scf_solvers.jl]: At least two spaces between code and `#` and one space after `#`

File: src/core/scf_solvers.jl [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    N = size(x0,1)
+    T = eltype(x0)
+    #xs: ring buffer storing the iterates, from newest to oldest
```
Comment (mfherbst) [inline src/core/scf_solvers.jl]: Spacing around `#`.

File: src/core/scf_solvers.jl [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        end
+
+        xs = circshift(xs,(0,1))
+        fs = circshift(fs,(0,1))
+        xs[:,1] = new_x
+    end
+    (sol=xs[:,1], converged=err < eps)
+end
+
+# CROP iterates maintain xn and fn (/!\ fn != f(xn)).
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/core/scf_solvers.jl]: Paper reference?

**Comment by mfherbst** [comment]
Another remark: I find it a bit strange that there is an "scf" function (inside src/core/scf.jl) and a "self_consistent_field" function (inside src/util/self_consistent_field.jl). I think we should merge them or rename them ...

**Comment by mfherbst** [comment]
For info: Currently the test fails for Float32 because the lobpcg_qr cannot achieve the tolerance of 1e-6 in Float32 precision.

**Comment by antoine-levitt** [comment]
Yup, asking for 1e-6 in Float32 is a bit ambitious

**Comment by mfherbst** [comment]
Rebased against master with new LOBPCG.

File: src/utils/self_consistent_field.jl [hunk 8]
```diff
@@ -13,22 +13,26 @@ for diagonalisation. Possible `algorithm`s are `:scf_nlsolve` or `:scf_damped`.
 function self_consistent_field(ham::Hamiltonian, n_bands::Int, n_electrons::Int;
                                ρ=nothing, tol=1e-6,
                                lobpcg_prec=PreconditionerKinetic(ham, α=0.1),
-                               max_iter=100, algorithm=:scf_nlsolve, kwargs...)
+                               max_iter=100, algorithm=:scf_nlsolve, damping=0.2, m=5, den_scaling = 0.0)
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/utils/self_consistent_field.jl]: Or actually I think I will merge this function and the core/scf, adopting a similar approach for selecting the `compute_occupation` functions as you did for the SCF solvers.

File: src/core/scf.jl [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    # Initialize guess wavefunctions if needed
+    if Psi == nothing
+        Psi = [Matrix(qr(randn(Complex{T}, length(pw.basis_wf[ik]), n_bands)).Q)
+               for ik in 1:length(pw.kpoints)]
+    end
+    values_hartree = empty_potential(ham.pot_hartree)
+    values_xc = empty_potential(ham.pot_xc)
+    energies = Dict{Symbol, real(eltype(ρ))}()
+    update_energies_potential!(energies, values_hartree, ham.pot_hartree, ρ)
+    update_energies_potential!(energies, values_xc, ham.pot_xc, ρ)
```
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/core/scf.jl]: Will be addressed during upcoming refactoring.

File: src/utils/self_consistent_field.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (mfherbst) [reply src/utils/self_consistent_field.jl]: Will be addressed during upcoming refactoring.

**Pull request merged**